README.md
# Validity90

This project aims on reverse engineering protocol of Validity 138a:0090, 138a:0094, 138a:0097, 06cb:0081 fingerprint readers, creating specification and FLOSS libfprint driver.

## Discussions

Main chat of this project: [![Gitter](https://img.shields.io/gitter/room/nwjs/nw.js.svg)](https://gitter.im/Validity90/Lobby?utm_source=share-link&utm_medium=link&utm_campaign=share-link)

libfrprint issue: [https://bugs.freedesktop.org/show_bug.cgi?id=94536](https://bugs.freedesktop.org/show_bug.cgi?id=94536)
Lenovo forums: [https://forums.lenovo.com/t5/Linux-Discussion/Validity-Fingerprint-Reader-Linux/td-p/3352145](https://forums.lenovo.com/t5/Linux-Discussion/Validity-Fingerprint-Reader-Linux/td-p/3352145)

## Notable files

1. [spec.md](spec.md) - Specification draft, the main work goes here right now.
2. [dissector.lua](dissector.lua) - Wireshark dissector for decrypting communication after key exchange.
3. ~[libfprint directory](libfprint) - libfprint repo with this driver integrated~ Not ready at the moment.
4. [prototype](prototype) - Standalone prototype(extremly ugly code, would be completly rewritten for driver)

## Testing

[Prototype](prototype) testers are needed, please share your result and join us in our [Gitter](https://gitter.im/Validity90/Lobby?utm_source=share-link&utm_medium=link&utm_campaign=share-link).

## Status

|   Device  | Status |
|-----------|--------|
| 138a:0090 | Scan, Image output, Internal DB check works in prototype |
| 138a:0097 | Scan, Internal DB check works in prototype |
| 138a:0094 | Doesn't work yet, but looks promising. I will try to work on it after 90/97 |
| 06cb:0081 | So far looks identical to 138a:0094 |
| 138a:0091 | Different protocol, out of scope for this project. |


| 		      Task       			| Specification/Analysis  | Prototype   | Driver 	    |
|---------------------------|-------------------------|-------------|-------------|
| Initialization  		      | Done 					          | Done	 	    | Not Started |
| Configuration/Reconfig    | In progress 	          | In progress | Not Started |
| Pre TLS key exchange 	    | In progress 				    | Done        | Not Started |
| TLS 			                | Done 						        | Done  	    | Not Started |
| 90: Ops: scan, LED, etc| In progress  			      | Scan, LED works  | Not Started |
| 97: Ops: enroll, check, reset, LED, etc| In progress  			      | Check works  | Not Started |
| Image format  		        | In progress  			      | Done        | Not Started |

## Donations

Please consider donating to support the development.  
With paypal: [![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/nmikhailov/0usd)  
With BTC: 1AsoCUGNVniBSBS359u2wKrA2CZiTgKZkM  
With ETH: 0x005D8d9b5609b9A3Ec8B3cBd463a949405A8d621  


libfprint/NEWS
This file lists notable changes in each release. For the full history of all
changes, see ChangeLog.

2017-05-14: v0.7.0 release
* Drivers:
  - Add VFS0050 driver
  - Fix possible crash in AES3500 and AES4000
  - Fix broken enrollment in VFS101
  - Better verification with small sensor scanners
  - Plenty of fixes in VFS5011
  - Fix memory corruption in AES1610
  - Improve calibration settings for AES1610
  - Improve image assembling in upeksonly driver
  - Autodetect whether image is encrypted in uru4k

* Library:
  - NBIS: Remove false minutia at the edge of partial image
  - Introduce routines to assemble image from lines (used in VFS5011 and upeksonly)
  - Fix a bug that can cause driver state machine to enter into endless loop. 

* Udev rules:
  - Add driver name to the USB properties

* Plenty of build fixes

2015-02-03: v0.6.0 release

* Drivers:
  - Reduce duplication between AES3500 and AES4000 drivers and
    add support for AES3500 device
  - Add support for UPEK 147e:2020 and Upek Eikon 2 devices
  - Add EgisTec ES603 driver
  - Add VFS5011 driver
  - Always perform 5 scans for image enrollment
  - Better verification with AES1660 driver
  - Better verification for a number of AES drivers

* Library:
  - Always use Pixman for image manipulation, gdk-pixbuf and ImageMagick
    are not supported any more.

* Udev rules:
  - Fix warning when USB hub or system does not support power management

2013-08-11: v0.5.1 release

* Drivers
  - Add support for 147e:2020 to upeke2 driver
  - Fix possible race condition, and cancellation in uru4000 driver

* Udev rules:
  - Add Microsoft keyboard to the suspend blacklist

* Plenty of build fixes

2012-12-03: v0.5.0 release

* Drivers:
  - New VFS300/VFS301 driver
  - New AES2550/AES2810 drivers
  - New AES1660 driver
  - New AES2660 driver
  - New DigitalPersona URU4500 driver
  - Avoid empty capture and improve image contrast in the
  AES2501 and AES2550 drivers
  - Update upektc driver, add support for Eikon Touch 300
  - Fix UrU4000 image capture on ARM

* Library:
  - Fix global variable collisions with libusb and other system headers
  - Fix possible crash in NBIS image processing with some fingerprints

* Udev rules:
  - Fix power control path for newer kernels
  - Move udev rules to the correct directory
  - Don't print duplicated udev rules
  - Include udev rules in the tarball
  - Allow disabling of udev rules for cross-compiling
  - Add driver names in the generated rules

2011-04-18: v0.4.0 release
 * Add support for Validity VFS101 (USB ID 138a:0001)
 * Fix crasher when resizing a fingerprint image
 * Fix wrong timeout being returned when either of
   libusb or libfprint doesn't have a timeout

2010-09-08: v0.3.0 release
 * Add support for UPEK TCS4C (USB ID 147e:1000)
 * Use NSS instead of OpenSSL for GPL compliance
 * upeksonly driver bug fixes
 * Fix a crash if a scan was shorter than 8 lines
 * Fix compilation with C++ compiler

2010-08-19: v0.2.0 release (since 0.1.0-pre2)
 * Add gdk-pixbuf support for image manipulation
 * Add udev rules to allow devices to autosuspend
 * Finish port of AES1610 driver
 * Add UPEK EikonII (TCRD4C) driver

2007-12-07: v0.0.5 release
 * AES1610 imaging improvements
 * Internal cleanups for Authentec drivers
 * Add support for latest Microsoft Fingerprint Scanner hardware revision

2007-11-22: v0.0.4 release
 * Enable AES1610 driver thanks to Michele B
 * Implement identification: one-to-many fingerprint matching (Daniel Drake)

2007-11-19: v0.0.3 release
 * Add API to access minutiae (Daniel Drake)
 * Add API to delete enroll data (Daniel Drake)
 * Add Authentec AES1610 driver (Anthony Bretaudeau)

2007-11-17: v0.0.2 release
 * Detect reversed scans on AES2501 (Vasily Khoruzhick)
 * Improved AES2501 scanning
 * Compatibility with older ImageMagick versions
 * Add UPEK TouchChip driver (Jan-Michael Brummer)
 * Add binarization API

2007-11-15: v0.0.1 release
 * Initial release



old/aes.py
from Crypto import Random
from Crypto.Cipher import AES
import base64
import hexdump


BLOCK_SIZE=32

def encrypt(message, passphrase):
    # passphrase MUST be 16, 24 or 32 bytes long, how can I do that ?
    IV = Random.new().read(BLOCK_SIZE)
    aes = AES.new(passphrase, AES.MODE_CFB, IV)
    return base64.b64encode(aes.encrypt(message))

#def decrypt(enc ):
    #enc = base64.b64decode(enc)
    #iv = enc[:16]
    #cipher = AES.new(self.key, AES.MODE_CBC, iv )
    #return unpad(cipher.decrypt( enc[16:] ))


k1 = bytearray.fromhex("6A56DB3E80F6DC0B")
k2 = bytearray.fromhex("74D6B42A391F3E05")
k3 = bytearray.fromhex("F51A60DA31DBC8C7")
k4 = bytearray.fromhex("BE40D4A499F13C68")

k1.reverse()
k2.reverse()
k3.reverse()
k4.reverse()

k1.extend(k2)
k1.extend(k3)
k1.extend(k4)

#data = bytearray.fromhex("d0d352deff4a42cd039842531a1d8962ae5b30ced3052b575ab782ac4a14cbae1461d63525f057e13f3c3b459eec327026788cb781bdda48d299f7586c234a40")


data = """
00000000: 40 71 BA 90 15 98 D0 1D  D4 47 27 56 59 D0 42 5A  @q.......G'VY.BZ
00000010: 68 02 3E A9 E0 AC 46 EC  EF D2 1B 7A E2 62 DD 77  h.>...F....z.b.w
00000020: ED 78 4D F6 3A DE 82 04  C2 B2 73 4A 01 16 42 37  .xM.:.....sJ..B7
00000030: 4A 51 AC 69 84 E0 6E AC  31 B8 04 41 F8 78 F4 13  JQ.i..n.1..A.x..
00000040: 9E 67 53 BC 6C 82 9F AB  49 D8 7F 64 00 98 3F 58  .gS.l...I..d..?X
00000050: 21 40 65 48 6D DB 6C 49  85 47 70 7E 94 29 C4 B8  !@eHm.lI.Gp~.)..
00000060: 57 79 00 2E B7 87 C9 C1  65 D7 CF 72 D8 7F 7D 4E  Wy......e..r..}N
"""

data ="""
0000: 33 5c ed 15 55 3c f5 f4 de 14 a0 f2 59 68 00 a2 
0010: a0 98 58 c2 06 67 d5 c1 06 e3 bf e6 6a ec 6a c0 
0020: 2d b2 d8 77 d9 0e c4 12 e3 ab 48 ab aa b4 b9 56 
0030: 75 30 69 9d 0a c3 d9 bb ff de 42 11 bd 34 03 21 
0040: cf a2 8d 3c 1b e4 ba f0 1f f4 40 69 6f b4 78 18 
0050: f3 2d 6b 22 80 86 64 31 14 34 2a 81 2c cc d7 c6 
0060: 62 f3 9e 5f 78 a6 39 d3 db 57 c3 30 d4 dd 12 8f 
"""

data = hexdump.restore(data)

#key = bytes(k1)
key = hexdump.restore("""
48 78 02 70 
5e 5a c4 a9 93 1c 44 aa 4d 32 25 22 39 e0 bf 8f 
0c 85 4d de 49 0c cc f6 87 ef ad 9c
""")
iv = bytes(data[:16])

cypher = AES.new(key, AES.MODE_CBC, iv)

print("key", key.hex())
print("data", data.hex())
print(cypher.decrypt(bytes(data[16:])).hex())
